### PR TITLE
Python console interactive help

### LINKED ISSUE #36338: Shortcut to open pyqgis API from QGIS console/script
**Feature description.**
It would be extremely useful to have a dynamic link from the python console/editor to the pyqgis API website.

Right now we have auto-completion (that is great) but typing something like F1 to open the browser at the page of the class/method would be wonderful.

### PR DESCRIPTION
**Pull request by YoannQDQ**
- Fix #36338

## Description

Add a shortcut in python editors to open the API

When pressing F1, open the API page matching
- selected text if any OR
- word at the mouse position (if enabled in settings, see below) OR
- word at the text cursor position

Behavior is slighltly different for the Python Console (and editors) and for the others generic QgsCodeEditorPython (e.g Expression Editor).

### Generic editors 
This PR only improves the already existing "Search in PyQGIS Documentation" action in the context menu to add the F1 shortcut and the word under mouse logic. Besides, if the text hovered matches a Qt class name, the Qt API is opened at the correct page. PyQGIS documentation urls cannot be deduced from the class name only as they also include the module name, so pressing F1 will only search the PyQGIS documentation.

### Python console (where the PR truly shines)

Add a widget in the devtools panel to display the API documentation (PyQGIS and Qt)

api_panel

The buttons at the top of the widget access the home pages of the APIs.
The button at the top right opens the current URL in the system web browser

When pressing F1 while hovering a symbol on the Python Console or Editors tab, the object is evaluated to get its class and module. The relevant API (PyQGIS or Qt) is then opened at the correct page. It works for class names, class, or even objects

Consider the following snippet:
```Python
myLayer = iface.activeLayer() 
if isinstance(myLayer, QgsRasterLayer):
  print("It is a QgsRasterLayer")
a = QLabel()
```
Let's say the active layer is a vector layer. With the text cursor at the following words, pressing F1 will
- `myLayer`: open the PyQGIS api page for QgsVectorLayer. 
- `iface`: open the PyQGIS api page for QgisInterface.
- QLabel: open the Qt api page for QLabel 
- QgsRasterLayer (in the print statement or in the type check): open the Qt api page for QgsRasterLayer 

If the symbol is not recognized, open the PyQGIS API home page. 

### Settings

context_help_settings

- **Open In**: how should the API html page be displayed?
  - Embedded Webview
  - Default system Web browser
- **F1 on Hover**: whether pressing F1 should trigger the context help for the hovered word or for the word at the textCursor

### Demo

console-interactive-help-demo

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice work @YoannQDQ !!

> which API should be displayed

Is there a compelling reason we'd want to use C++ now? In the last month the PyQGIS docs.have greatly improved, and I'd argue they are at least as good as the C++ ones now...

> PyQt doc is often incomplete)

This is the right call for the qt docs 👍

**Comment by uclaros** [comment]
This is great!
I find the _hover + F1_ functionality a little counter intuitive, though.
I would expect the word at text cursor position to always take precedence.

**Comment by YoannQDQ** [comment]
> Is there a compelling reason we'd want to use C++ now? In the last month the PyQGIS docs.have greatly improved, and I'd argue they are at least as good as the C++ ones now...

Out of habit, I mostly read the C++ API doc myself, but I'm completely willing to letting it go. Should it also be removed from the help menu?

> I find the hover + F1 functionality a little counter intuitive, though

I took it from QtCreator, but if it doesn't feel right, we can ditch it. I'd appreciate some more feedback on this.

**Comment by uclaros** [comment]
> I took it from QtCreator, but if it doesn't feel right, we can ditch it. I'd appreciate some more feedback on this.

Just curious, what QtCreator version is this in? On my 13.0.2 the hovered item is ignored, and it makes perfect sense for me as the mouse cursor may be _anywhere_ when navigating with the keyboard!

**Comment by YoannQDQ** [comment]
It's an **ancient** version of qt creator that we keep using on a specific project for ungodly reasons... Namely QtCreator 4.11. This behavior has indeed disappeared in later versions.

**Comment by YoannQDQ** [comment]
@nyalldawson The "search" function of the PyQGIS documentation works with QWebEngineView. Are there any plan to make it available in C++? Maybe integrate it to QgsWebView?

**Comment by nyalldawson** [comment]
@YoannQDQ 

We should just be able to use the QWebEngineView class directly, without any need for a Qgs wrapper class. We'd just need to conditionally use it only in `#ifdef HAVE_WEBENGINE` blocks

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -13125,6 +13112,96 @@ void QgisApp::unregisterDevToolFactory( QgsDevToolWidgetFactory *factory )
   mDevToolFactories.removeAll( factory );
 }
 
+
+void QgisApp::showApiDocumentation( const QString &api, bool embedded, const QString &object, const QString &module )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:13116]: Could you move the guts of this to QgsDocumentationPanelWidget ? I'd like to reduce the amount of code in qgisapp.cpp, which is already ridiculously large

File: src/ui/qgsdocumentationpanelbase.ui [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  </property>
+  <property name="windowIcon">
+   <iconset resource="../../images/images.qrc">
+    <normaloff>:/images/themes/default/mActionHelpContents.svg</normaloff>:/images/themes/default/mActionHelpContents.svg</iconset>
+  </property>
+  <layout class="QVBoxLayout" name="verticalLayout">
+   <item>
+    <layout class="QHBoxLayout" name="horizontalLayout">
+     <item>
+      <widget class="QToolButton" name="mPyQgisHomeButton">
```
Comment (nyalldawson) [inline src/ui/qgsdocumentationpanelbase.ui:24]: Let's set these buttons to autoraise true -- it'll clean up the panel appearance a little.

File: src/gui/qgisinterface.h [hunk 3]
```diff
@@ -1280,6 +1280,16 @@ class GUI_EXPORT QgisInterface : public QObject
...
 
+    /**
+     * Show a page of the API documentation
+     * \param api "pyqgis" or "qgis" or "qt" or "pyqgis-search"
+     * \param embedded If TRUE, the documentation will be opened in the embedded devtools webview. Otherwise, use system web browser
+     * \param object object to show in the documentation
+     * \param module used only if api = "pyqgis"
+     * \since QGIS 3.42
+    */
+    virtual void showApiDocumentation( const QString &api = QStringLiteral( "pyqgis" ), bool embedded = true, const QString &object = QString(), const QString &module = QString() ) = 0;
```
Comment (nyalldawson) [inline src/gui/qgisinterface.h:1291]: Instead of a raw string for `api`, can you add an enum class into the Qgis class for this? 

I would also change the boolean "embedded" to an explicit enum class too, eg:

```
enum class DocumentationBrowser
{
   DeveloperToolsPanel,
   SystemWebBrowser
}
```

This makes the code more readable (and slightly more future proof) vs a raw boolean

File: src/gui/codeeditors/qgscodeeditorpython.h [hunk 4]
```diff
@@ -129,6 +130,13 @@ class GUI_EXPORT QgsCodeEditorPython : public QgsCodeEditor
      */
     void searchSelectedTextInPyQGISDocs();
 
+    /**
+     * Searches the given text in the official APIs (PyQGIS, C++ QGIS or Qt) documentation.
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.h:134]:
```suggestion
     * Displays the given text in the official APIs (PyQGIS, C++ QGIS or Qt) documentation.
```

File: src/app/qgsdevtoolspanelwidget.h [hunk 5]
```diff
@@ -32,19 +33,23 @@ class APP_EXPORT QgsDevToolsPanelWidget : public QWidget, private Ui::QgsDevTool
...
+    void showUrl( const QUrl &url );
+
   private slots:
 
     void setCurrentTool( int row );
 
   private:
 
     QMap< QgsDevToolWidgetFactory *, int> mFactoryPages;
+    QgsDocumentationPanelWidget *mDocumentationPanel;
```
Comment (nyalldawson) [inline src/app/qgsdevtoolspanelwidget.h:52]:
```suggestion
    QgsDocumentationPanelWidget *mDocumentationPanel = nullptr;
```

File: src/app/qgsdevtoolspanelwidget.cpp [hunk 6]
```diff
@@ -43,6 +49,21 @@ QgsDevToolsPanelWidget::QgsDevToolsPanelWidget( const QList<QgsDevToolWidgetFact
 
 QgsDevToolsPanelWidget::~QgsDevToolsPanelWidget() = default;
 
+void QgsDevToolsPanelWidget::addToolWidget( QgsDevToolWidget *widget )
+{
+  mStackedWidget->addWidget( widget );
+
+  QListWidgetItem *item = new QListWidgetItem( widget->windowIcon(), QString() );
+  item->setToolTip( widget->windowTitle() );
+  item->setData( Qt::UserRole, widget->windowTitle() );
```
Comment (nyalldawson) [inline src/app/qgsdevtoolspanelwidget.cpp:58]:
```suggestion
  item->setData( Qt::UserRole, widget->objectName() );
```

Otherwise we're potentially using a translated value, and won't be able to hardcode in an identifier for the documentation browser item (see below)

File: src/app/qgsdevtoolspanelwidget.cpp [hunk 7]
```diff
@@ -103,3 +124,9 @@ void QgsDevToolsPanelWidget::setCurrentTool( int row )
   whileBlocking( mOptionsListWidget )->setCurrentRow( row );
   mStackedWidget->setCurrentIndex( row );
 }
+
+void QgsDevToolsPanelWidget::showUrl( const QUrl &url )
+{
+  whileBlocking( mOptionsListWidget )->setCurrentRow( 0 );
```
Comment (nyalldawson) [inline src/app/qgsdevtoolspanelwidget.cpp:130]: This is a little fragile, eg if we move around the rows later. I think instead you should call setActiveTab here.

File: python/console/console_settings.ui [hunk 8]
```diff
@@ -466,116 +583,46 @@
...
+            <item>
+             <property name="text">
+              <string>Embedded Webview (developer tools)</string>
+             </property>
+            </item>
+            <item>
+             <property name="text">
+              <string>Default system web browser</string>
+             </property>
+            </item>
```
Comment (nyalldawson) [inline python/console/console_settings.ui:619]: Can you populate via code instead, and use user data for each item? Eg this approach: https://github.com/qgis/QGIS/blob/master/src/gui/layout/qgslayoutscalebarwidget.cpp#L132

(It's more robust vs populating in the ui)

**Comment by YoannQDQ** [comment]
Thanks for the review! I also added the conditional QWebEngineView include.

**Comment by YoannQDQ** [comment]
Didn't manage to make the fedora 39 build work with QWebEngineView... Any clue @nyalldawson?

**Comment by YoannQDQ** [comment]
@nyalldawson Would it be possible to merge this now?

**Comment by nyalldawson** [comment]
@YoannQDQ 
>  Would it be possible to merge this now?

Sorry, was off last week with FOSS4G Oceania ... it's going to take me a bit of time to catch back up on QGIS work now :laughing:

File: CMakeLists.txt [hunk 9]
```diff
@@ -557,13 +557,7 @@ if(WITH_CORE)
...
   set (WITH_QTWEBENGINE FALSE CACHE BOOL "Enable QtWebEngine support")
   if (WITH_QTWEBENGINE)
-    if(BUILD_WITH_QT6)
-      find_package(${QT_VERSION_BASE} COMPONENTS WebEngineCore REQUIRED)
-    else()
-      # QWebEnginePage is only available in widgets module in Qt5
-      find_package(${QT_VERSION_BASE} COMPONENTS WebEngineCore WebEngineWidgets REQUIRED)
-    endif()
-
+    find_package(${QT_VERSION_BASE} COMPONENTS WebEngineCore WebEngineWidgets REQUIRED)
```
Comment (nyalldawson) [inline CMakeLists.txt:560]: Is this change required? I was hoping to avoid the widget library when building without GUI. Could the WebEngineWidgets find package be moved to src/gui/CMakeLists.txt?

File: src/gui/codeeditors/qgscodeeditorpython.h [hunk 10]
```diff
@@ -120,6 +122,13 @@ class GUI_EXPORT QgsCodeEditorPython : public QgsCodeEditor
 
     bool checkSyntax() override;
 
+    /**
+     *  Checks  whether QWebEngineView is available to display context help
+     *
+     * \since QGIS 3.42
+     */
+    static bool hasWebEngine();
```
Comment (nyalldawson) [inline src/gui/codeeditors/qgscodeeditorpython.h:130]: Can you move this to QgsGui ? It's not specifically related to QgsCodeEditorPython, and I think having it in a more generic place would be valuable.

File: src/core/qgis.h [hunk 11]
```diff
@@ -5573,6 +5573,32 @@ class CORE_EXPORT Qgis
     };
     Q_ENUM( ColorModel )
 
+    /**
+     * Documentation Api
```
Comment (nyalldawson) [inline src/core/qgis.h:5577]:
```suggestion
     * Documentation API
```

File: src/core/CMakeLists.txt [hunk 12]
```diff
@@ -2612,17 +2612,10 @@ if (WITH_PDAL)
...
-    else()
-      # QWebEnginePage is in WebEngineWidgets for Qt5
-      target_link_libraries(qgis_core
-        ${QT_VERSION_BASE}::WebEngineCore
-        ${QT_VERSION_BASE}::WebEngineWidgets
-      )
-    endif()
+  target_link_libraries(qgis_core
+    ${QT_VERSION_BASE}::WebEngineCore
+    ${QT_VERSION_BASE}::WebEngineWidgets
```
Comment (nyalldawson) [inline src/core/CMakeLists.txt:2617]: as above -- can we try to avoid the widget import where we can (ie qt 6), and only do it for the gui library?

File: python/console/console_settings.py [hunk 13]
```diff
@@ -76,6 +76,10 @@ def __init__(self, parent):
         self.parent = parent
         self.setupUi(self)
 
+        # Populate the documentation Browser combobox
+        self.contextHelpBrowser.addItem(QCoreApplication.translate("PythonConsole", "Embedded Webview (developer tools)"), Qgis.DocumentationBrowser.DeveloperToolsPanel)
+        self.contextHelpBrowser.addItem(QCoreApplication.translate("PythonConsole", "Default system web browser"), Qgis.DocumentationBrowser.SystemWebBrowser)
```
Comment (nyalldawson) [inline python/console/console_settings.py:81]:
```suggestion
        self.contextHelpBrowser.addItem(QCoreApplication.translate("PythonConsole", "Embedded Webview (developer tools)"), Qgis.DocumentationBrowser.DeveloperToolsPanel)
        self.contextHelpBrowser.addItem(QCoreApplication.translate("PythonConsole", "Default System Web Browser"), Qgis.DocumentationBrowser.SystemWebBrowser)
```

File: src/gui/qgsgui.h [hunk 14]
```diff
@@ -318,6 +318,13 @@ class GUI_EXPORT QgsGui : public QObject
      */
     static void initCalloutWidgets() SIP_SKIP;
 
+    /**
+     *  Checks  whether QWebEngineView is available to display context help
```
Comment (nyalldawson) [inline src/gui/qgsgui.h:322]:
```suggestion
     *  Checks whether QWebEngineView is available to display HTML content.
```

File: python/PyQt6/gui/auto_generated/qgsgui.sip.in [hunk 15]
```diff
@@ -251,6 +251,13 @@ Returns the screen at the given global ``point`` (pixel).
 
 
 
+    static bool hasWebEngine();
+%Docstring
+Checks  whether QWebEngineView is available to display context help
```
Comment (nyalldawson) [inline python/PyQt6/gui/auto_generated/qgsgui.sip.in:256]:
```suggestion
Checks whether QWebEngineView is available to display HTML content.
```

File: python/gui/auto_generated/qgsgui.sip.in [hunk 16]
```diff
@@ -251,6 +251,13 @@ Returns the screen at the given global ``point`` (pixel).
 
 
 
+    static bool hasWebEngine();
+%Docstring
+Checks  whether QWebEngineView is available to display context help
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsgui.sip.in:256]:
```suggestion
Checks whether QWebEngineView is available to display HTML content.
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thanks @YoannQDQ !

**Comment by GallPeters** [comment]
Great feature! Is it possible to use local documentation in the embedded widget (for offline use)? I tried changing the link that points to the documentation in the general settings but it did not affect the widget. 

By the way, when accessing the qgis api docs either in the embedded widget or in the browser, a link opens that points to the documentation of version 3.42 that does not exist (needs to be changed to latest).

The option to enter a url manually will solve both issues - entering a link to local documentation or to selected documentation of the desired version.

**Comment by DelazJ** [comment]
@YoannQDQ Pressing F1 key doesn't seem to work correctly on Windows. See #62337

**Comment by YoannQDQ** [comment]
Indeed. The shortcut is overshadowed by the default Help Content Shortcut.

As a workaround, you can disabled the shortcut in the KeyBoard Shortcuts editor but it is rather annoying. 

image


image

I guess a real fix would involve handling `QEvent::ShortcutOverride` in `QgsCodeEditor::event`
